logstash-core/lib/logstash/instrument/collector.rb

```diff
@@ -1,79 +1,109 @@
+    #
+    # @param [Time] Time of execution
+    # @param [result] Result of the execution
+    # @param [Exception] Exception
+    def update(time_of_execution, result, exception)
+      return true if exception.nil?
+      logger.error("Collector: Something went wrong went sending data to the observers", 
+                   :execution_time => time_of_execution,
+                   :result => result,
+                   :exception => exception)
     end
 
-    private
-    def roll_over?
-      Concurrent.monotonic_time - @last_rotation >= self.class.snapshot_rotation_time
+    # Snapshot the current Metric Store and return it immediately,
+    # This is useful if you want to get acecess to the current metric store without
+    # waiting for a periodic call.
+    # 
+    # @return [LogStash::Instrument::MetricStore]
+    def snapshot_metric
+      @metric_store.dup
     end
 
-    def snapshot
-      # TODO: We currently sent delta to the observers.
-      # Should we keep the whole picture and send the updated state instead?
-      if roll_over? && @snapshot_rotation_mutex.try_lock 
-        # fair rotation of the snapshot done by the winning thread
-        # metric could be written in the previous snapshot.
-        # Since the snapshot isn't written right away
-        # the view of the snapshot should be consistent at the time of
-        # writing, if we don't receive any events for 5 secs we wont send it.
-        # This might be a problem, for time correlation.
-        logger.debug("Collector: Rotating snapshot", :last_rotation => @last_rotation) if logger.debug?
-
-        publish_snapshot
-        rotate_snapshot
-
-        @snapshot_rotation_mutex.unlock
-      end
-
-      @current_snapshot
+    private
+    # Use the string to generate a concrete class for this metrics
+    # 
+    # @param [String] The name of the class
+    # @raise
+    def concrete_class(type)
+      LogStash::Instrument::MetricType.const_get(type.capitalize)
     end
 
-    def rotate_snapshot
-      @current_snapshot = Snapshot.new
-      update_last_rotation
+    # Configure and start the periodic task for snapshotting the `MetricStore`
+    def start_periodic_snapshotting
+      @snapshot_task = Concurrent::TimerTask.new { publish_snapshot }
+      @snapshot_task.execution_interval = SNAPSHOT_ROTATION_TIME
+      @snapshot_task.timeout_interval = SNAPSHOT_ROTATION_TIME_TIMEOUT
+      @snapshot_task.add_observer(self)
+      @snapshot_task.execute
     end
 
+    # Create a snapshot of the MetricStore and send it to to the registered observers
+    # The observer will receive the following signature in the update methode.
+    #
+    # `#update(created_at, metric_store)`
     def publish_snapshot
-      changed
-      notify_observers(Concurrent.monotonic_time, @current_snapshot) 
-    end
-
-    def update_last_rotation
-      @last_rotation = Concurrent.monotonic_time
+      created_at = Time.now
```

Comment:
Time.now is faster, I assumed Concurrurent.monotonic_time was faster and this seems to be wrong.

https://github.com/elastic/logstash/pull/4358#issuecomment-165324127